### ISSUE TITLE
Panics in procedural macros can lead to hundreds of unrelated resolution errors

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#89362]
One thing I've often seen with rustc when I'm developing a procedural macro is that I'm in a situation where a procedural macro is generating a bunch of code and I've got a file that's using all the code the procedural macro. Everything works well but inevitably I've got a bug in the macro and the macro panics. This causes the macro to not actually produce any items, but all the code I have that was hand-written and using the results of the generated macro is still there. This leads rustc to generate an error-per-use-of-generated code, which often for me at least can be in the dozens or hundreds of errors. I find that I basically have to sift through all the compiler errors back to the top to find the one with the procedural macro to actually find the issue.

Ideally I think rustc would recognize macro expansion errors and stop producing other name resolution errors related to the module that had the macro expansion error. Given that the macro could have possibly produced anything else in the module any other further error could be a false positive.

As a small example with this procedural macro:

```rust
// pm.rs
extern crate proc_macro;

use proc_macro::TokenStream;

#[proc_macro]
pub fn foo(a: TokenStream) -> TokenStream {
    if a.to_string() == "a" {
        panic!()
    }
    "pub fn foo() {}".parse().unwrap()
}
```

and a `main.rs`:

```rust
// main.rs
pm::foo!();

fn main() {
    foo();
}
```

compiling this yields no error:

```
$ rustc pm.rs --crate-type proc-macro
$ rustc main.rs --extern pm=./libpm.so
```

but changing `main.rs` to this:

```rust
// main.rs
pm::foo!(a);

fn main() {
    foo();
}
```

yields:

```
$ rustc pm.rs --crate-type proc-macro
$ rustc main.rs --extern pm=./libpm.so
error: proc macro panicked
 --> main.rs:1:1
  |
1 | pm::foo!(a);
  | ^^^^^^^^^^^^
  |
  = help: message: explicit panic

error[E0425]: cannot find function `foo` in this scope
 --> main.rs:4:5
  |
4 |     foo();
  |     ^^^ not found in this scope

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0425`.
```

The second error here isn't very helpful because it's just a knock-on effect of the first error. While this example itself is quite isolated and easy to debug, it's a bit more troublesome to sift through hundreds of knock-on effects from a mistake in a procedural macro than one.
